kasa/xortransport.py

```diff
@@ -0,0 +1,225 @@
+    class.
+    """
+
+    DEFAULT_PORT: int = 9999
+    BLOCK_SIZE = 4
+
+    def __init__(self, *, config: DeviceConfig) -> None:
+        super().__init__(config=config)
+        self.reader: Optional[asyncio.StreamReader] = None
+        self.writer: Optional[asyncio.StreamWriter] = None
+        self.query_lock = asyncio.Lock()
+        self.loop: Optional[asyncio.AbstractEventLoop] = None
+
+    @property
+    def default_port(self):
+        """Default port for the transport."""
+        return self.DEFAULT_PORT
+
+    @property
+    def credentials_hash(self) -> str:
+        """The hashed credentials used by the transport."""
+        return ""
+
+    async def _connect(self, timeout: int) -> None:
+        """Try to connect or reconnect to the device."""
+        if self.writer:
+            return
+        self.reader = self.writer = None
+
+        task = asyncio.open_connection(self._host, self._port)
+        async with asyncio_timeout(timeout):
+            self.reader, self.writer = await task
+            sock: socket.socket = self.writer.get_extra_info("socket")
+            # Ensure our packets get sent without delay as we do all
+            # our writes in a single go and we do not want any buffering
+            # which would needlessly delay the request or risk overloading
+            # the buffer on the device
+            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+
+    async def _execute_send(self, request: str) -> Dict:
+        """Execute a query on the device and wait for the response."""
+        assert self.writer is not None  # noqa: S101
+        assert self.reader is not None  # noqa: S101
+        debug_log = _LOGGER.isEnabledFor(logging.DEBUG)
+        if debug_log:
+            _LOGGER.debug("%s >> %s", self._host, request)
+        self.writer.write(XorEncryption.encrypt(request))
+        await self.writer.drain()
+
+        packed_block_size = await self.reader.readexactly(self.BLOCK_SIZE)
+        length = _UNSIGNED_INT_NETWORK_ORDER.unpack(packed_block_size)[0]
+
+        buffer = await self.reader.readexactly(length)
+        response = XorEncryption.decrypt(buffer)
+        json_payload = json_loads(response)
+        if debug_log:
+            _LOGGER.debug("%s << %s", self._host, pf(json_payload))
+
+        return json_payload
+
+    async def close(self) -> None:
+        """Close the connection."""
+        writer = self.writer
+        self.close_without_wait()
+        if writer:
+            with contextlib.suppress(Exception):
+                await writer.wait_closed()
+
+    def close_without_wait(self) -> None:
+        """Close the connection without waiting for the connection to close."""
+        writer = self.writer
+        self.reader = self.writer = None
+        if writer:
+            writer.close()
+
+    async def reset(self) -> None:
+        """Reset the transport."""
+        await self.close()
```

Comment:
```suggestion
```

I this we should drop `reset` here since there is no way to reset the state given the protocol is stateful and we don't have a way to tell the device to reset